fetch/cosmwasm_contract/src/migration.rs

```diff
@@ -0,0 +1,29 @@
+use crate::access_control::wipe_all_roles_from_storage;
+use crate::contract::initialise_contract_state;
+use crate::error::ERR_ALREADY_REFUNDED;
+use crate::msg::MigrateMsg;
+use crate::state::is_state_valid;
+use cosmwasm_std::{entry_point, DepsMut, Env, Response, StdError, StdResult};
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, env: Env, msg: MigrateMsg) -> StdResult<Response> {
+    // Remove all previously registered roles
+    wipe_all_roles_from_storage(deps.storage);
+
+    // Ensure correct contract state
+    if let Some(re_init) = msg.re_init {
+        initialise_contract_state(
+            deps.storage,
+            &env,
+            &re_init.admin,
+            re_init.supply,
+            re_init.relay_eon,
+            re_init.fees_accrued,
```

Comment:
These 3 attributes should be simply read from the current contract state (= no change in their value).
Though we can query those values from the contract and before initialisation and use them in this re_init message, but it kind of feels dangerous, opening the gate to possible mishaps (wrong copy-paste, ors omething like that).

Reply:
We are be able to read it from contract state only if original contract state is compatible with new contract state or if we keep the original contract state definition for migration sake.